File: 백준/Gold/11657. 타임머신/타임머신.java
import java.io.IOException;
import java.util.List;
import java.util.ArrayList;

public class Main {
    private static final long INFINITY = Long.MAX_VALUE;
    public static void main(String[] args) throws IOException {
        int N = readInt(), M = readInt();
        Graph graph = Graph.getInstance(N);
        for (int i = 0; i < M; i++) {
            graph.addEdge(readInt(), readInt(), readInt());
        }
        BellmanFord bellmanFord = BellmanFord.getInstance(graph, 1);
        long[] distances = bellmanFord.getShortestDistances();
        StringBuilder print = new StringBuilder();
        for (int i = 2; i <= N; i++) {
            if (distances[i] == INFINITY) {
                print.append("-1");
            } else print.append(distances[i]);
            print.append("\n");
        }
        if (isSourceIsolated(distances)) {
            System.out.println(print);
            return;
        }
        if (bellmanFord.hasNegativeWeightedCycle()) {
            System.out.println(-1);
            return;
        }
        System.out.println(print);
    }
    private static boolean isSourceIsolated(long[] distances) {
        if (distances.length == 1) return false;
        for (int i = 2; i < distances.length; i++) {
            if (distances[i] != INFINITY) return false;
        }
        return true;
    }
    private static int readInt() throws IOException {
        int val = 0;
        int c = System.in.read();
        while (c <= ' ') {
            c = System.in.read();
        }
        boolean minus = false;
        if (c == '-') {
            minus = true;
            c = System.in.read();
        }
        do {
            val = 10 * val + c - 48;
        } while ((c = System.in.read()) >= 48 && c <= 57);
        if (minus) return -val;
        return val;
    }
}

class BellmanFord {
    private Graph graph;
    private long[] distances;
    private static final long INFINITY = Long.MAX_VALUE;

    private BellmanFord(Graph graph, int source) {
        this.graph = graph;
        int n = graph.getVertexCount();
        this.distances = new long[n + 1];
        for (int i = 1; i <= n; i++) {
            if (i != source) distances[i] = INFINITY;
        }
        relaxWeights();
    }

    public static BellmanFord getInstance(Graph graph, int source) {
        return new BellmanFord(graph, source);
    }

    private void relaxWeights() {
        int n = graph.getVertexCount();
        List<Edge> edges = graph.getEdges();
        for (int i = 0; i < n; i++) {
            for (Edge edge : edges) {
                if (distances[edge.getSource()] == INFINITY) continue;
                if (distances[edge.getDestination()] > distances[edge.getSource()] + edge.getWeight()) {
                    distances[edge.getDestination()] = distances[edge.getSource()] + edge.getWeight();
                }
            }
        }
    }

    public long[] getShortestDistances() {
        return this.distances;
    }

    public boolean hasNegativeWeightedCycle() {
        for (Edge edge : graph.getEdges()) {
            if (distances[edge.getDestination()] > distances[edge.getSource()] + edge.getWeight()) {
                return true;
            }
        }
        return false;
    }
}

class Graph {
    private final int vertexCount;
    private List<Edge> edges;

    private Graph(int n) {
        this.vertexCount = n;
        edges = new ArrayList<>();
    }

    public static Graph getInstance(int n) {
        return new Graph(n);
    }

    public int getVertexCount() {
        return vertexCount;
    }

    public List<Edge> getEdges() {
        return edges;
    }

    public void addEdge(int source, int destination, long weight) {
        edges.add(Edge.getInstance(source, destination, weight));
    }
}

class Edge {
    private final int source;
    private final int destination;
    private final long weight;

    private Edge(int source, int destination, long weight) {
        this.source = source;
        this.destination = destination;
        this.weight = weight;
    }

    public static Edge getInstance(int source, int destination, long weight) {
        return new Edge(source, destination, weight);
    }

    public int getSource() {
        return source;
    }

    public int getDestination() {
        return destination;
    }

    public long getWeight() {
        return weight;
    }
}

File: 백준/Gold/11657. 타임머신/타임머신.cc
#include <iostream>
#include <vector>
#include <limits>
using namespace std;

const long long INF = numeric_limits<long long>::max();

struct Edge {
    int source;
    int destination;
    long long weight;
    Edge(int s, int d, long long w) : source(s), destination(d), weight(w) {}
};

class Graph {
public:
    int vertexCount;
    vector<Edge> edges;

    Graph(int n) : vertexCount(n) { }

    void addEdge(int source, int destination, long long weight) {
        edges.emplace_back(source, destination, weight);
    }
};

class BellmanFord {
private:
    Graph &graph;
    vector<long long> distances; // 1-indexed: distances[1] ~ distances[vertexCount]
public:
    BellmanFord(Graph &g, int source) : graph(g) {
        distances.resize(graph.vertexCount + 1, INF);
        // 초기: source 정점의 거리는 0, 나머지는 INF
        distances[source] = 0;
        relaxWeights();
    }

    void relaxWeights() {
        int n = graph.vertexCount;
        // n-1번 반복해도 충분하지만, Java 코드에서는 n번 반복합니다.
        for (int i = 0; i < n; i++) {
            for (const Edge &edge : graph.edges) {
                if (distances[edge.source] == INF) continue;
                if (distances[edge.destination] > distances[edge.source] + edge.weight) {
                    distances[edge.destination] = distances[edge.source] + edge.weight;
                }
            }
        }
    }

    const vector<long long>& getShortestDistances() const {
        return distances;
    }

    bool hasNegativeWeightedCycle() {
        // 한 번 더 relaxation 시도하여 변경되면 음수 사이클이 존재하는 것으로 판단
        for (const Edge &edge : graph.edges) {
            if (distances[edge.source] == INF) continue;
            if (distances[edge.destination] > distances[edge.source] + edge.weight) {
                return true;
            }
        }
        return false;
    }
};

bool isSourceIsolated(const vector<long long>& distances) {
    // 정점의 개수가 1이면 고립된 것으로 보지 않음.
    if (distances.size() <= 2) return false;
    // 2번 정점부터 마지막 정점까지 모두 INF면, source(1번 정점)는 다른 정점과 연결되어 있지 않은 것으로 판단.
    for (size_t i = 2; i < distances.size(); i++) {
        if (distances[i] != INF) return false;
    }
    return true;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M;
    cin >> N >> M;
    Graph graph(N);
    for (int i = 0; i < M; i++) {
        int a, b;
        long long t;
        cin >> a >> b >> t;
        // 주어진 Java 코드에서는 정점 번호가 그대로 사용됩니다.
        graph.addEdge(a, b, t);
    }

    // source는 1번 정점
    BellmanFord bellmanFord(graph, 1);
    const vector<long long>& distances = bellmanFord.getShortestDistances();

    // 결과를 저장할 문자열(또는 버퍼) 대신, 출력할 때 바로 처리합니다.
    // 2번 정점부터 N번 정점까지의 최단 거리를 출력합니다.
    // 만약 INF이면 -1을 출력합니다.
    bool sourceIsolated = isSourceIsolated(distances);
    if (sourceIsolated) {
        for (int i = 2; i <= N; i++) {
            cout << (distances[i] == INF ? -1 : distances[i]) << "\n";
        }
        return 0;
    }

    if (bellmanFord.hasNegativeWeightedCycle()) {
        cout << -1 << "\n";
        return 0;
    }

    for (int i = 2; i <= N; i++) {
        cout << (distances[i] == INF ? -1 : distances[i]) << "\n";
    }
    return 0;
}